\documentclass[a0,plainsections,30pt]{sciposter}

\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{amsthm}
\usepackage{amssymb}
\usepackage{bbm}
\usepackage{natbib}

\setlength{\columnseprule}{0pt}

\newcommand{\Expect}{\mathbb{E}}
\newcommand{\etaopt}{\eta^{*}}

\newcommand{\targetexpectation}{\Expect_{q_{\eta^*}}
\left[\#\{\substack{\text{distinct clusters}\\\text{in new dataset}}\} \right]}

\usepackage[framemethod=TikZ, xcolor=RGB]{mdframed}

\definecolor{mydarkblue}{rgb}{0,.06,.5}
\definecolor{mydarkred}{rgb}{.5,0,.1}
\definecolor{myroyalblue}{rgb}{0,.1,.8}

\usepackage{sectsty}
\sectionfont{\color{mydarkblue}\centering\LARGE\bf}

\mdfdefinestyle{MyFrame}{%
    linecolor=mydarkblue,
    outerlinewidth=2pt,
    roundcorner=20pt,
    innertopmargin=10pt,
    innerbottommargin=10pt,
    innerrightmargin=10pt,
    innerleftmargin=10pt,
    backgroundcolor=blue!10}


\title{\textcolor{mydarkblue}{
Evaluating Sensitivity to the Stick Breaking Prior in Bayesian Nonparametrics
}}

\author{%
Runjing Liu\textsuperscript{1*} \quad
Ryan Giordano\textsuperscript{2*} \quad
Michael I. Jordan\textsuperscript{1} \quad
Tamara Broderick\textsuperscript{2} \\
{\large\normalfont
 \textsuperscript{1}Department of Statistics, UC Berkeley\quad
 \textsuperscript{2}CSAIL, MIT\quad
 \textsuperscript{*}Equal contribution authors}
}

\leftlogo[1]{static_images/logo_left.png}
\rightlogo[1]{static_images/faces.jpg}

% Fiddle with the margin
\addtolength{\topmargin}{-0.5in}
\addtolength{\textheight}{1in}
\begin{document}
\conference{BNP 2019}

\setlength{\parskip}{0.25em}

\maketitle

\vspace{-1in}

<<initialization, echo=FALSE, message=FALSE, results='asis'>>=
library(tidyverse)
knitr_debug <- FALSE
source("./R_graphs/Initialize.R")

# Common graphs library
source("./R_graphs/CommonGraphs.R")

# This file is produced by R_graphs/data/iris_data/make_knitr_dataset.R.
iris_data <- LoadIntoEnvironment(
    file.path(data_path, "iris_data/iris_data_for_knitr.Rdata"))

# Puts genomics_data in the environment.
source("./R_graphs/ProcessGenomicsData.R")

# Turn off caching if you need to regenerate any of the R code.
knitr_debug <- FALSE # Set to true to see error output
cache_knitr <- TRUE # Set to true to cache knitr output for this analysis.
@


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% FIRST COLUMN
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{minipage}[t]{0.45\textwidth}

\begin{mdframed}[style=MyFrame]
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%
\section*{Overview}
\vspace{-0.3in}
%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{itemize}
\item Researchers often want to estimate the number of distinct clusters
that would be present in a new dataset.

\item A variational Bayes (VB) approximation to a Bayesian nonparametric (BNP)
model makes inferring the number of clusters amenable to Bayesian inference.
%We approximate the exact posterior with variational Bayes (VB).

\item \textbf{Question}: How sensitive are the
VB approximation and the resulting inference to BNP model choices?

\item \textbf{Problem}: Re-running VB for multiple model choices is expensive.

\item \textbf{We propose}: A local approximation to efficiently
estimate BNP sensitivity from a single run of VB, avoiding
expensive refitting.

%\item We evaluate sensitivity to parametric and functional perturbations.
\end{itemize}
\end{mdframed}
\vspace{-0.7in}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Model and inference }
\vspace{-0.3in}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

The \textbf{Dirichlet process} is a popular Bayesian nonparametric
(BNP) model used for clustering.  A stick-breaking representation uses
a distribution on sticks $\{\nu_k\}$ to induce a prior on component
probabilities $\{\pi_k\}$.

\begin{figure}
\centering
\includegraphics[width = 0.95\textwidth]{./static_images/DP_stick_breaking.png}
\caption{The Dirichlet stick-breaking process with $\nu_k \sim \mathrm{Beta}(1, \alpha)$.}
\end{figure}
%
\vspace{-0.3in}
Define:
\begin{itemize}
\item $\epsilon :=$ Any real-valued hyperparameter for the stick-breaking prior
    $p(\nu_k | \epsilon)$.
\item $\etaopt :=$ The optimal variational parameters.
\item $\targetexpectation :=$ The quantity we want to infer.
\end{itemize}

\begin{mdframed}[style=MyFrame]
VB inference defines a map:
%
\begin{gather*}
\textrm{Stick-breaking prior}
    \mapsto \textrm{VB approximation}
    \mapsto \textrm{Posterior expected \# clusters},\\
\textrm{i.e.,}\\
\epsilon
    \mapsto \etaopt
    \mapsto \targetexpectation.
\end{gather*}
\end{mdframed}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%

\textbf{Parametric perturbation.} To stay within the class of Dirichlet
processes, we might take $\epsilon := \alpha - \alpha_0$ and $p(\nu_k | \alpha) :=
\mathrm{Beta}(\nu_k | 1, \alpha)$.

\textbf{Functional perturbation.} Suppose we are interested in the effect of
replacing the original beta prior $p_0(\nu_k) := \mathrm{Beta}(\nu_k | 1,
\alpha)$ with another distribution $p_1(\nu_k)$. Then we take $\epsilon :=
\delta$ in
%
\vspace{-0.3in}
\begin{align*}
p(\nu_k \vert \delta) \propto
    p_{0}(\nu_k)\left(\frac{p_1(\nu_k)}{p_0(\nu_k)}\right)^\delta.
\end{align*}
\vspace{-0.3in}

By using a multiplicative perturbation, the map $p_1 \mapsto \etaopt$ is
Fr\'{e}chet differentiable in the norm $\left\Vert p_1 / p_0
\right\Vert_\infty$.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%
\vspace{-0.3in}
\subsection*{Local approximation to dependence on the prior}
\vspace{-0.2in}
%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Let $\epsilon=0$ represent the original prior at which we fit a VB
approximation.

\begin{mdframed}[style=MyFrame]
For any other $\epsilon \ne 0$ we approximate $\epsilon \mapsto \etaopt$ with
\begin{align*}
\eta^*(\epsilon)  &\approx  \eta^*(0) +
\frac{d \eta^*(\epsilon)}{d\epsilon^T}\Big|_{\epsilon=0} \epsilon.
\end{align*}
We retain the non-linearities in $\etaopt \mapsto \targetexpectation$.
\end{mdframed}

\begin{itemize}
\item We can \textbf{automatically compute}
    the local approximation
    using automatic differentiation \cite{giordano:2017:covariances}.
    %\cite{giordano:2017:covariances, maclaurin:2015:autograd}.
\item The approximation requires the
    \textbf{inverse Hessian of the VB objective}.
\item This is an application of \textbf{local Bayesian robustness}
\citep{gustafson:1996:localposterior}.
\end{itemize}

\end{minipage}
\hfill \vrule \hfill
\begin{minipage}[t]{0.45\textwidth}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% SECOND COLUMN
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%
\section*{Data}
\vspace{-0.3in}
%
\begin{minipage}[t]{0.49\textwidth}
%
\textbf{Iris data.}
For our first dataset, we discard species
information and cluster the 150 four-dimensional observations of the Fisher iris
dataset \citep{iris_data_anderson}.
The loadings on the first two principle components are shown
to the right.
%
\end{minipage}
%
\begin{minipage}[t]{0.49\textwidth}
<<iris_pca_cap, cache=FALSE>>=
SetImageSize(aspect_ratio=2.0 * base_aspect_ratio,
             image_width=0.45 * base_image_width)
@
%
\vspace{-0.4in}
<<iris_pca, cache=cache_knitr, fig.show='hold'>>=
ggplot(iris_data$iris_df) +
  geom_point(aes(x=PC1, y=PC2))
@
\end{minipage}
%

\vspace{1em}

\begin{minipage}[t]{0.49\textwidth}
%
\textbf{Mouse data.} For our second dataset, we use a publicly available
dataset of gene expression in mice \citep{shoemaker:2015:ultrasensitive}. We
smooth and center 7000 gene expression time series and cluster the resulting
patterns.  A typical smoothed time series is shown to the right.  The faint
lines show regression posterior uncertainty.
%
\end{minipage}
%
\begin{minipage}[t]{0.49\textwidth}
\vspace{-0.8in}
\begin{figure}[!h]
\centering
\includegraphics[width = 0.95\textwidth]{./static_images/mouse_genes.png}
\end{figure}
\end{minipage}
%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%
\vspace{-0.6in}
\section*{Results}
\vspace{-0.3in}

\textbf{Parametric perturbation.} Vertical lines show the location of
$\alpha_0$.
\begin{itemize}
\item VB + BNP is robust for the mouse data but not the iris data.
\item The approximation \textbf{extrapolates best from more to fewer clusters.}
\end{itemize}

\vspace{0.05in}
%
<<fig_cap1, cache=FALSE>>=
fig_cap1 <- "Re-optimization vs approximation."
SetImageSize(aspect_ratio=1.0 * base_aspect_ratio)
@
<<param_sens_plot, cache=cache_knitr, fig.show='hold'>>=
results_df_1 <- filter(iris_data$results_df, !pred, pert=="alpha3")
results_df_2 <- filter(iris_data$results_df, !pred, pert=="alpha8")
results_df_3 <- filter(iris_data$results_df, !pred, pert=="alpha13")
grid.arrange(
    plot_parametric_sensitivity(results_df_1, alpha_0 = 3.0) +
      ggtitle('Iris data') +
        theme(legend.position = 'none', legend.title=element_blank()),
    plot_parametric_sensitivity(results_df_2, alpha_0 = 8.0) +
      ggtitle(' ') + theme(legend.position="none"),
    plot_parametric_sensitivity(results_df_3, alpha_0 = 13.0) +
      ggtitle(' ') + theme(legend.position="none"),
    ncol  = 3)
@

<<gene_param_sens_plot, cache=cache_knitr, fig.show='hold'>>=
grid.arrange(
    plot_parametric_sensitivity(
      genomics_data$processed_results %>%
        filter(!inflate, !alpha_increase, !functional),
      alpha_0=2.0) +
      ggtitle("Mouse data")  + theme(legend.position="none"),
  plot_parametric_sensitivity(
    genomics_data$processed_results %>%
      filter(!inflate, alpha_increase, !functional),
    alpha_0=2.0) +
    ggtitle(" ") + theme(legend.position=c(0.75, 0.65)),
    ncol=2
)
@

\textbf{Functional perturbation.} The first row shows the original prior
$p_0(\nu_k)$ and the functional perturbation $p_1(\nu_k)$.  Again, VB + BNP
is robust for the mouse data but not the iris data, and the approximation
extrapolates best from more to fewer clusters.

\vspace{0.1in}
<<fig_cap2, cache=FALSE>>=
SetImageSize(aspect_ratio= 2.0 * base_aspect_ratio)
@
<<functional_sens_plot, cache=cache_knitr, fig.show='hold'>>=
iris_pert1 <- filter(iris_data$prior_pert_df, filename == "prior_pert1.csv")
iris_pert2 <- filter(iris_data$prior_pert_df, filename == "prior_pert2.csv")
iris_pred_pert1 <- filter(iris_data$results_df, pred, pert=="1")
iris_pred_pert2 <- filter(iris_data$results_df, pred, pert=="2")

gene_pred_pert <- filter(genomics_data$processed_results,
                         functional, !inflate)

grid.arrange(
    # First row --- include titles.
    plot_prior_perturbation(iris_pert1) +
        theme(legend.position = "None") +
        ggtitle(TeX("\\textbf{Iris data, first $p_1$}")),

    plot_prior_perturbation(iris_pert2) +
        theme(legend.position = "None") +
        ggtitle(TeX("\\textbf{Iris data, second $p_1$}")),

    plot_prior_perturbation(genomics_data$processed_pert_df) +
        theme(legend.position = c(0.45, 0.80)) +
        ggtitle("Mouse data"),

    # Second row --- results
    plot_parametric_sensitivity(iris_pred_pert1, xlabel=TeX("$\\delta$")) +
        theme(legend.position = "None"),

    plot_parametric_sensitivity(iris_pred_pert2, xlabel=TeX("$\\delta$")) +
        theme(legend.position =  "None"),

    plot_parametric_sensitivity(gene_pred_pert, xlabel=TeX("$\\delta$")) +
        theme(legend.position =  c(0.7, 0.8)),
    ncol=3
)
@

\noindent\rule{0.95\textwidth}{1pt}

\textbf{Contact and code:} rgiordan@mit.edu, runjing\_liu@berkeley.edu
{\color{blue} https://github.com/Runjing-Liu120/sensitivity\_to\_stick\_breaking\_in\_BNP}
{\color{blue} https://github.com/rgiordan/vittles}
(sensitivity analysis)

\end{minipage}\\

\begin{center}
\noindent\rule{0.95\textwidth}{1pt}
\end{center}
\input{acknowledgements}

\renewcommand{\section}[2]{}%
\footnotesize{
  \bibliographystyle{abbrv}
  \bibliography{./references}
}

\end{document}
